src/core/tools/contextValidator.ts

```diff
@@ -0,0 +1,170 @@
+
+/**
+ * More aggressive buffer percentage specifically for file reading validation.
+ * This is separate from the global TOKEN_BUFFER_PERCENTAGE to provide extra safety
+ * when reading files without affecting other context window calculations.
+ */
+const FILE_READ_BUFFER_PERCENTAGE = 0.25 // 25% buffer for file reads
+
+export interface ContextValidationResult {
+	shouldLimit: boolean
+	safeMaxLines: number
+	reason?: string
+}
+
+/**
+ * Validates if a file can be safely read based on its size and current runtime context state.
+ * Reads lines incrementally and counts tokens as it goes, stopping when reaching the token limit.
+ * Returns a safe maxReadFileLine value to prevent context overflow.
+ */
+export async function validateFileSizeForContext(
+	filePath: string,
+	totalLines: number,
+	currentMaxReadFileLine: number,
+	cline: Task,
+): Promise<ContextValidationResult> {
+	try {
+		// Get actual runtime state from the task
+		const modelInfo = cline.api.getModel().info
+		const { contextTokens: currentContextTokens } = cline.getTokenUsage()
+		const contextWindow = modelInfo.contextWindow
+
+		// Get the model-specific max output tokens using the same logic as sliding window
+		const modelId = cline.api.getModel().id
+		const apiProvider = cline.apiConfiguration.apiProvider
+		const settings = await cline.providerRef.deref()?.getState()
+
+		// Map apiProvider to the format expected by getModelMaxOutputTokens
+		let format: "anthropic" | "openai" | "gemini" | "openrouter" | undefined
+		if (
+			apiProvider === "anthropic" ||
+			apiProvider === "bedrock" ||
+			apiProvider === "vertex" ||
+			apiProvider === "claude-code"
+		) {
+			format = "anthropic"
+		} else if (apiProvider === "openrouter") {
+			format = "openrouter"
+		} else if (apiProvider === "openai" || apiProvider === "openai-native") {
+			format = "openai"
+		} else if (apiProvider === "gemini" || apiProvider === "gemini-cli") {
+			format = "gemini"
+		}
+
+		const maxResponseTokens = getModelMaxOutputTokens({ modelId, model: modelInfo, settings, format })
+
+		// Calculate how much context is already used
+		const currentlyUsed = currentContextTokens || 0
+
+		// Calculate remaining context space
+		const remainingContext = contextWindow - currentlyUsed
+
+		// Apply buffer to the remaining context, not the total context window
+		// This gives us a more accurate assessment of what's actually available
+		const usableRemainingContext = Math.floor(remainingContext * (1 - FILE_READ_BUFFER_PERCENTAGE))
+
+		// Use the same approach as sliding window: reserve the model's max tokens
+		// This ensures consistency across the codebase
+		const reservedForResponse = maxResponseTokens || 0
+
+		// Calculate available tokens for file content
+		const availableTokensForFile = usableRemainingContext - reservedForResponse
+
+		// Now read lines incrementally and count tokens until we reach the limit
+		const BATCH_SIZE = 100 // Read 100 lines at a time
```

Comment:
Is there a scenario where this file read patterns is slow? I'm thinking about the case where `countTokens` is an API call and the file has a ton of short or empty lines. It seems like we want to trigger a read after some data threshold has been reached as opposed to a line count. Thoughts?